action/protocol/execution/evm/evmstatedbadapter.go

```diff
@@ -690,7 +709,12 @@ func (stateDB *StateDBAdapter) getNewContract(addr hash.Hash160) (Contract, erro
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to load account state for address %x", addr)
 	}
-	contract, err := newContract(addr, account, stateDB.dao, stateDB.cb)
+	var contract Contract
+	if stateDB.saveHistory {
+		contract, err = newContract(addr, account, stateDB.dao, stateDB.cb, HistoryRetentionOption(stateDB.blockHeight))
```

Comment:
Extracting dao and cb from sm is not a good way of using sm. I am trying to find a good way to use sm directly in newContract. This is also the reason why you need to pass in saveHistory.